### ISSUE TITLE
Snippets

### ISSUE THREAD

**Issue by bbslipers** [SpartanJ/ecode#111]
Everything is just great.
The only thing missing is snippets.

**Comment by SpartanJ** [comment]
I haven't found many LSPs supporting snippets. Are you requesting built-in snippets for things like loops and stuffs or do you want to be able to define custom snippets? What language are you using? Can you provide more information on how would you want it to behave or some demonstration using other editor? I think this is something it can be added without much work, but I never used them so I'm not very familiar on how people tent to use them. Thanks

**Comment by bbslipers** [comment]
I use snippets to insert my own code snippets. I am completely satisfied with the snippet system in the Vim SnipMate plugin. But to be completely honest, the most convenient implementation of the snippet system that I have met is present in the RJ TextEd editor.

![image](https://github.com/SpartanJ/ecode/assets/20872380/691496ba-d55a-4f27-8570-9508c645e07e)

**Comment by zigsby** [comment]
Yep, snippets would be great. I am very used to snippets from SlickEdit. Here is roughly how they work there:

- The user can define their own sections of text, as short or as long as they please. In SlickEdit they can be set up on a per language basis. So e.g. a snippet for seeding a new file could always have the same name (alias), but there would actually be different text for different languages.
- Each snippet is associated with an 'alias'. That is a code word to be typed in the source code. When the space bar is pressed with the cursor located just after the alias and the alias being the last piece of text in the line, the snippet will be inserted. So snippet insertion is in the flow of typing. If the alias happens to be embedded in text (not at the end of a line) then a hotkey press is required.
- Snippets can have embedded keys to be expanded upon insertion. For example a date, time, file name and various other things can be automatically expanded upon insertion. 
- Snippets can have embedded variables. If a variable exists in the snippet, a dialog will open upon insertion to allow typing in values for each variable. These will then be replaced in the snippet before insertion.
- Snippets can contain control codes to control behavior. E.g. move the cursor to the start of the line or to a specific column before insertion, place the cursor at a specific location when done with the insertion, etc.

I use this extensively for starting new files, the insertion of function headers, if/else control blocks, switch/case, loops, comment clocks, etc. etc.

**Comment by zigsby** [comment]
BTW, Cuda Text offers similar functionality as well. It's actually quite common now.

**Comment by SpartanJ** [comment]
It's still not designed, so I don't know.

**Comment by ghost** [comment]
Linked: https://github.com/vala-lang/vala-language-server/issues/342

**Comment by SpartanJ** [comment]
LSP snippets support has been added, now it's possible to navigate over tab-stops and has full [LSP snippet syntax](https://microsoft.github.io/language-server-protocol/specifications/lsp/3.18/specification/#snippet_syntax) support.

**Comment by SpartanJ** [comment]
Added user-defined snippets based on vscode snippets (also supports vscode snippets out of the box), documentation [here](https://github.com/SpartanJ/ecode/blob/develop/docs/snippets.md).

**Comment by ghost** [comment]
Please add option to not auto import snippets from `.vscode` dir. I don't like this behavior.

**Comment by SpartanJ** [comment]
Loading snippets from `.vscode` is enabled by default. To disable it, set `load_vscode_snippets` to `false` in the autocomplete plugin's configuration (`autocomplete.json` file). Snippets from `.ecode` and the user snippet directory remain enabled.

**Comment by ghost** [comment]
> Loading snippets from `.vscode` is enabled by default. To disable it, set `load_vscode_snippets` to `false` in the autocomplete plugin's configuration (`autocomplete.json` file). Snippets from `.ecode` and the user snippet directory remain enabled.

Let me link the specific commit here for reference: https://github.com/SpartanJ/eepp/commit/da4d201eebc97f2834497003279f4ecc8846ee71. Thank you.
